packages/region/src/services/region-module.ts

```diff
@@ -115,20 +119,65 @@ export default class RegionModuleService<
       sharedContext
     )
 
-    let currencyMap = new Map(currencies.map((c) => [c.code.toLowerCase(), c]))
-    for (const reg of data) {
-      const lowerCasedCurrency = reg.currency_code.toLowerCase()
+    const countriesInDb = await this.countryService_.list(
+      { iso_2: data.map((d) => d.countries).flat() },
+      { select: ["iso_2", "region_id"] },
+      sharedContext
+    )
+
+    const countriesInDbMap = new Map(countriesInDb.map((c) => [c.iso_2, c]))
+
+    const currencyMap = new Map(
+      currencies.map((c) => [c.code.toLowerCase(), c])
+    )
+
+    const toCreate: InternalCreateRegionDTO[] = []
+    for (const region of data) {
+      const reg = { ...region } as InternalCreateRegionDTO
+      const countriesToAdd = region.countries || []
+
+      if (countriesToAdd) {
+        const notInDb = countriesToAdd.filter(
+          (c) => !countriesInDbMap.has(c.toLowerCase())
+        )
+
+        if (notInDb.length) {
+          throw new MedusaError(
+            MedusaError.Types.INVALID_DATA,
+            `Countries with codes ${countriesToAdd.join(", ")} does not exist`
+          )
+        }
+
+        const regionCountries = countriesToAdd.map((code) => {
+          const country = countriesInDbMap.get(code.toLowerCase())
+
+          if (country?.region_id) {
+            throw new MedusaError(
+              MedusaError.Types.INVALID_DATA,
+              `Country with code ${code} is already assigned to a region`
+            )
+          }
+
+          return country
+        }) as unknown as TCountry[]
```

Comment:
suggestion: split the mapping of countries to db entities from the validation of region_id for readability. 

We're not doing any db queries and everything is constant operation so the operation would still be linear in cost (though 2N instead of just N): 

```
 const alreadyHasRegion = countriesToAdd.filter(
          (c) => countriesInDbMap.get(c.toLowerCase())?.region_id
        )

        if (alreadyHasRegion.length) {
          throw new MedusaError(
            MedusaError.Types.INVALID_DATA,
            `Country with code ${alreadyHasRegion[0]} is already assigned to a region`
          )
        }

        reg.countries = countriesToAdd.map(
          (code) => countriesInDbMap.get(code.toLowerCase())!
        )
```